src/modalities/nn/moe.py

```diff
@@ -0,0 +1,141 @@
+from typing import Callable, Optional, Tuple
+
+import torch
+import torch.nn as nn
+from pydantic import BaseModel
+
+
+# MoE implementation inspired from Dbrx https://github.com/databricks/dbrx/blob/main/model/modeling_dbrx.py
+class MoEFFNConfig(BaseModel):
+    moe_num_experts: int
+    moe_top_k: int
+    moe_normalize_expert_weights: float
+    uniform_expert_assignment: bool
+    ffn_hidden_size: int
+    act_fn: Callable[[], nn.Module] = nn.SiLU
+    moe_jitter_eps: Optional[float]
+
+
+class MoERouter(nn.Module):
+    def __init__(self, hidden_size: int, moe_config: MoEFFNConfig):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.moe_num_experts = moe_config.moe_num_experts
+        self.moe_top_k = moe_config.moe_top_k
+        self.moe_normalize_expert_weights = moe_config.moe_normalize_expert_weights
+        self.uniform_expert_assignment = moe_config.uniform_expert_assignment
+        self.moe_jitter_eps = moe_config.moe_jitter_eps
+
+        self.layer = nn.Linear(self.hidden_size, self.moe_num_experts, bias=False)
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.LongTensor]:
+        if self.training and self.moe_jitter_eps is not None:
+            x = x * self._jitter(x)
+
+        weights = self.layer(x.view(-1, x.shape[-1])).softmax(dim=-1, dtype=torch.float32)
+        top_weights, top_experts = torch.topk(weights, self.moe_top_k, dim=-1)
+
+        if self.moe_normalize_expert_weights:
+            top_weights = top_weights / torch.norm(
+                top_weights, p=self.moe_normalize_expert_weights, dim=-1, keepdim=True
+            )
+
+        if self.uniform_expert_assignment:
```

Comment:
then the top_k experts are chosen uniformly but the returned weights stay as they are, right?

in which cases would we do that?

Reply:
Danke danke. As for you question, Exactly, if the uniform_expert_assignment is set to True, indeed the experts are chosen uniformly,  this approach in my understanding is useful for load balancing ( there are no favorite experts, favorite experts can both hider the training of other experts and also create a bottleneck on themselves ).  Setting this flag to True though does sort of ignore the router features, so probably isnt best practice. I wonder if combining the two settings makes sense, initially have expert assignment be uniform and only later when the experts have had chance to learn this flag is switched to False.

Reply:
Okay, thanks for clarifying. I guess the initialization in random / uniformly anyways. In the configs the value is set to True. Is that intended?

Reply:
Thanks for pointing this out, For initial phase we wanted to setup the basic architecture of experts so this was kept as True, We are testing/working on adding the dynamic routing part.